cli/src/main/java/com/devonfw/tools/ide/tool/PluginBasedCommandlet.java

```diff
@@ -0,0 +1,134 @@
+
+  private Collection<PluginDescriptor> configuredPlugins;
+
+  /**
+   * The constructor.
+   *
+   * @param context the {@link IdeContext}.
+   * @param tool the {@link #getName() tool name}.
+   * @param tags the {@link #getTags() tags} classifying the tool. Should be created via {@link Set#of(Object) Set.of}
+   * method.
+   */
+  public PluginBasedCommandlet(IdeContext context, String tool, Set<String> tags) {
+
+    super(context, tool, tags);
+  }
+
+
+  protected Map<String, PluginDescriptor> getPluginsMap() {
+
+    if (this.pluginsMap == null) {
+      Map<String, PluginDescriptor> map = new HashMap<>();
+
+      // Load project-specific plugins
+      Path pluginsPath = getPluginsConfigPath();
+      loadPluginsFromDirectory(map, pluginsPath);
+
+      // Load user-specific plugins
+      Path userPluginsPath = getUserHomePluginsConfigPath();
+      loadPluginsFromDirectory(map, userPluginsPath);
+
+      this.pluginsMap = map;
+    }
+
+    return this.pluginsMap;
+  }
+
+  private void loadPluginsFromDirectory(Map<String, PluginDescriptor> map, Path pluginsPath) {
+    if (Files.isDirectory(pluginsPath)) {
+      try (Stream<Path> childStream = Files.list(pluginsPath)) {
+        Iterator<Path> iterator = childStream.iterator();
+        while (iterator.hasNext()) {
+          Path child = iterator.next();
+          String filename = child.getFileName().toString();
+          if (filename.endsWith(IdeContext.EXT_PROPERTIES) && Files.exists(child)) {
+            PluginDescriptor descriptor = PluginDescriptorImpl.of(child, this.context, isPluginUrlNeeded());
+
+            // Priority to user-specific plugins
+            map.put(descriptor.getName(), descriptor);
+          }
+        }
+      } catch (IOException e) {
+        throw new IllegalStateException("Failed to list children of directory " + pluginsPath, e);
+      }
+    }
+  }
+
+  /**
+   * @return {@code true} if {@link PluginDescriptor#getUrl() plugin URL} property is needed, {@code false} otherwise.
+   */
+  protected boolean isPluginUrlNeeded() {
+
+    return false;
+  }
+
+  protected Path getPluginsConfigPath() {
+
+    return this.context.getSettingsPath().resolve(this.tool).resolve(IdeContext.FOLDER_PLUGINS);
+  }
+
+  private Path getUserHomePluginsConfigPath() {
+
+    return Paths.get(System.getProperty("user.home"), ".ide", "settings", this.tool, IdeContext.FOLDER_PLUGINS);
+  }
+
+
+  /**
+   * @return the immutable {@link Collection} of {@link PluginDescriptor}s configured for this IDE tool.
+   */
+  public Collection<PluginDescriptor> getConfiguredPlugins() {
+
+    if (this.configuredPlugins == null) {
+      this.configuredPlugins = Collections.unmodifiableCollection(getPluginsMap().values());
+    }
+    return this.configuredPlugins;
+  }
+
+  /**
+   * @return the {@link Path} where the plugins of this {@link IdeToolCommandlet} shall be installed.
+   */
+  public Path getPluginsInstallationPath() {
+
+    // TODO add edition???
```

Comment:
This is a good question!
However, our current approach in devonfw-ide was:
* when we install or update a tool, we simply delete the plugins folder for that tool and reinstall all plugins from scratch. Please also note that IntelliJ supports getting the plugins based on the IntelliJ version so we might get different plugin binaries for different IntelliJ versions. For this reason we do not keep plugins but simply delete and reinstall them when the product changes.
* we can only have one installation of a tool inside a devonfw-ide instance. Therefore we can not have multiple editions of the same tool in the same instance and there is no conflict.